### PR TITLE
feat: Wynn Font functions

### PR DESCRIPTION
**Pull request by Rafii2198**
Adds new WynnFontModel that works akin to WynnAlphabetModel but for fonts and functions related to that. Thanks to them people don't need to have some external tools to create them, or spend hours doing them manually, but also as these are functions, they are dynamic which is just impossible or extremely hard to do manually

`to_fancy_text()` will change the text to fancy one
image

`to_background_text()` will change the text to the one with the background
image

### CODE REVIEW

**Comment by Rafii2198** [comment]
I noticed that it jumps a bit, I dunno if it is something inevitable or something that can be fixed, if so I would require a bit of help to do that
java_3CBvCVFAjA

File: common/src/main/java/com/wynntils/models/wynnfont/WynnFontModel.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+            'v', 'w', 'x', 'y', 'z', '?', '[', ']', '\\', '%', '&', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9');
+    private static final List<Character> fancyCharacters = List.of(
+            '\uE040', '\uE041', '\uE042', '\uE043', '\uE044', '\uE045', '\uE046', '\uE047', '\uE048', '\uE049',
+            '\uE04A', '\uE04B', '\uE04C', '\uE04D', '\uE04E', '\uE04F', '\uE050', '\uE051', '\uE052', '\uE053',
+            '\uE054', '\uE055', '\uE056', '\uE057', '\uE058', '\uE059', '\uE05A', '\uE05B', '\uE05C', '\uE05D',
+            '\uE05E', '\uE05F', '\uE060', '\uE061', '\uE062', '\uE063', '\uE064', '\uE065', '\uE066', '\uE067',
+            '\uE068', '\uE069');
+    private static final Map<Character, Character> normalToFancy = new HashMap<>();
+
+    private void createFontMaps() {
```
Comment (ryanzhoudev) [inline common/src/main/java/com/wynntils/models/wynnfont/WynnFontModel.java]: I think Map.of() would be better here. It'll give us concrete bindings from char->fancy instead of having to rely on index matching

**Comment by magicus** [comment]
The jumping seems to be due to the font being proportional-width and not fixed-width. I don't know if you can do anything about that.

File: common/src/main/java/com/wynntils/models/wynnfont/WynnFontModel.java (hunk 1 above)
Comment (magicus) [reply common/src/main/java/com/wynntils/models/wynnfont/WynnFontModel.java]: It might be too many items to fit in the limited Map.of() constructor. If so, a static block adding the content to a map is an alternative.

File: common/src/main/resources/assets/wynntils/lang/en_us.json [hunk 2]
```diff
@@ -2331,6 +2331,16 @@
...
   "function.wynntils.timeSinceLastKill.description": "The time, in milliseconds, since you last killed a mob",
   "function.wynntils.timeSinceLastKill.name": "Time Since Last Kill",
+  "function.wynntils.toBackgroundText.argument.backgroundColor": "Color of the background",
+  "function.wynntils.toBackgroundText.argument.leftEdge": "Style of the left edge. NONE or PILL",
+  "function.wynntils.toBackgroundText.argument.rightEdge": "Style of the right edge. NONE or PILL",
+  "function.wynntils.toBackgroundText.argument.text": "Text to convert. Incorrect characters will be omtited",
+  "function.wynntils.toBackgroundText.argument.textColor": "Color of the text",
+  "function.wynntils.toBackgroundText.description": "Converts the text to use Wynncraft background style.",
+  "function.wynntils.toBackgroundText.name": "To Background Text",
+  "function.wynntils.toFancyText.argument.text": "Text to convert. Incorrect characters will be omtited",
```
Comment (magicus) [inline common/src/main/resources/assets/wynntils/lang/en_us.json]: Typo. Also, I don't think the word "incorrect" is proper here. Better perhaps with "unsupported". (Also on line 2337).

**Comment by magicus** [comment]
Nice! I am just wondering about the word "fancy". Maybe call it "small-caps style"?

**Comment by Rafii2198** [comment]
> The jumping seems to be due to the font being proportional-width and not fixed-width. I don't know if you can do anything about that.

I believe they also have a different size of negative spaces and background characters to make it more smooth, I am just extremely bad at finding them, so I think it could be done, I just don't know the symbols for that different negative space or if that exist. Finding the negative space for the first edge took me like a half of an hour, and I only found them by copying someone else's work.

> Nice! I am just wondering about the word "fancy". Maybe call it "small-caps style"?

In files, it's actually called `five.png` so to make it more official, I can rename it to `toFiveFont`, the main reason I called it fancy is that on Discord when someone refers to that font they usually call it "fancy", probably because of lack of a better word.

**Comment by magicus** [comment]
I don't think "Five Font" is helpful to anyone. Sounds like an internal Wynncraft code name. Better to keep it as "Fancy" in that case.

**Comment by Rafii2198** [comment]
Perhaps we should do a pool on discord where people can vote for the name, I don't know personally which one would be the best. Or ask wynncraft devs how they call it so we can adapt the official name if there is any.

**Comment by kristofbolyai** [comment]
I'd suggest renaming to fancy font and merge.

File: common/src/main/java/com/wynntils/functions/WynnFontFunctions.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+package com.wynntils.functions;
+
+import com.wynntils.core.components.Models;
+import com.wynntils.core.consumers.functions.Function;
+import com.wynntils.core.consumers.functions.arguments.FunctionArguments;
+import com.wynntils.utils.colors.CustomColor;
+import java.util.List;
+
+public class WynnFontFunctions {
+    public static class toFancyTextFunction extends Function<String> {
```
Comment (ShadowCat117) [inline common/src/main/java/com/wynntils/functions/WynnFontFunctions.java]:
```suggestion
    public static class ToFancyTextFunction extends Function<String> {
```

File: common/src/main/java/com/wynntils/functions/WynnFontFunctions.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        }
+
+        @Override
+        public FunctionArguments.Builder getArgumentsBuilder() {
+            return new FunctionArguments.RequiredArgumentBuilder(
+                    List.of(new FunctionArguments.Argument<>("text", String.class, null)));
+        }
+    }
+
+    public static class toBackgroundTextFunction extends Function<String> {
```
Comment (ShadowCat117) [inline common/src/main/java/com/wynntils/functions/WynnFontFunctions.java]:
```suggestion
    public static class ToBackgroundTextFunction extends Function<String> {
```

File: common/src/main/java/com/wynntils/functions/WynnFontFunctions.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        }
+
+        @Override
+        public FunctionArguments.Builder getArgumentsBuilder() {
+            return new FunctionArguments.RequiredArgumentBuilder(List.of(
+                    new FunctionArguments.Argument<>("text", String.class, null),
+                    new FunctionArguments.Argument<>("textColor", CustomColor.class, null),
+                    new FunctionArguments.Argument<>("backgroundColor", CustomColor.class, null),
+                    new FunctionArguments.Argument<>("leftEdge", String.class, null),
+                    new FunctionArguments.Argument<>("rightEdge", String.class, null)));
```
Comment (ShadowCat117) [inline common/src/main/java/com/wynntils/functions/WynnFontFunctions.java:48]: Maybe set None as default for these?

File: common/src/main/java/com/wynntils/models/wynnfont/BackgroundEdge.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        this.edgeType = edgeType;
+        this.characterPair = characterPair;
+    }
+
+    public static BackgroundEdge fromString(String type) {
+        try {
+            return valueOf(type.toUpperCase(Locale.ROOT));
+        } catch (IllegalArgumentException e) {
+            return null;
+        }
```
Comment (ShadowCat117) [inline common/src/main/java/com/wynntils/models/wynnfont/BackgroundEdge.java]: Assuming None is a reasonable default

```suggestion
        for (BackgroundEdge edge : values()) {
            if (edge.edgeType.equalsIgnoreCase(type)) {
                return edge;
            }
        }

        return NONE;
```

File: common/src/main/java/com/wynntils/models/wynnfont/WynnFontModel.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+ */
+package com.wynntils.models.wynnfont;
+
+import com.wynntils.core.components.Model;
+import com.wynntils.utils.colors.CustomColor;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+
+public class WynnFontModel extends Model {
```
Comment (ShadowCat117) [inline common/src/main/java/com/wynntils/models/wynnfont/WynnFontModel.java]:
```suggestion
public final class WynnFontModel extends Model {
```

File: common/src/main/java/com/wynntils/models/wynnfont/WynnFontModel.java [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    private static final char NEGATIVE_SPACE = '\uE012';
+    private static final char NEGATIVE_SPACE_EDGE = '\u2064';
+    private static final char BACKGROUND = '\uE00F';
+    private static final Map<Character, Character> normalToFancy = Map.ofEntries(
```
Comment (ShadowCat117) [inline common/src/main/java/com/wynntils/models/wynnfont/WynnFontModel.java]: There was a comment for some you're missing here https://discord.com/channels/394189072635133952/1257255724061425674/1390188363318825051

**Review by ShadowCat117 — CHANGES_REQUESTED** [review]
NONE BackgroundEdge either needs a non null characterPair or you need to handle it in the model

```
Exception thrown by com.wynntils.functions.WynnFontFunctions$toBackgroundTextFunction
 java.lang.NullPointerException: Cannot invoke "com.wynntils.utils.type.Pair.a()" because "this.characterPair" is null
```

File: common/src/main/java/com/wynntils/functions/WynnFontFunctions.java (hunk 5 above)
Comment (Rafii2198) [reply common/src/main/java/com/wynntils/functions/WynnFontFunctions.java:48]: This doesn't matter as you need to provide all arguments of the function anyway

Comment (ShadowCat117) [reply common/src/main/java/com/wynntils/functions/WynnFontFunctions.java:48]: Oh can we not provide defaults for reguired arguments, that's a shame

File: common/src/main/java/com/wynntils/models/wynnfont/WynnFontModel.java (hunk 8 above)
Comment (Rafii2198) [reply common/src/main/java/com/wynntils/models/wynnfont/WynnFontModel.java]: I will need to do a quick test, from what I saw previously is that they are different width, and it makes it so they like overflow in background text, if that's the case then I will have to split the map of normalToFancy into 2

Comment (Rafii2198) [reply common/src/main/java/com/wynntils/models/wynnfont/WynnFontModel.java]: Yeah there is something wrong at the ends where it like places the background character at the end, covering the normal edge and leaving a bit of space
image

**Comment by ShadowCat117** [comment]
What if you generated the maps in a similar way to `WynnAlphabetModel`, so the characters that work on both can be generated for both the fancy and background maps, then the characters supported by the fancy font can be added to that map?

File: common/src/main/java/com/wynntils/models/wynnfont/WynnFontModel.java [hunk 9]
```diff
@@ -14,101 +15,30 @@ public final class WynnFontModel extends Model {
...
+    public static final List<Character> normalCharacters = List.of(
+            'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u',
+            'v', 'w', 'x', 'y', 'z', '?', '[', ']', '\\', '%', '&', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
+            '!', '(', ')', '<', '=', '>');
+    private static final List<Character> fancyCharacters = List.of(
+            '\uE040', '\uE041', '\uE042', '\uE043', '\uE044', '\uE045', '\uE046', '\uE047', '\uE048', '\uE049',
+            '\uE04A', '\uE04B', '\uE04C', '\uE04D', '\uE04E', '\uE04F', '\uE050', '\uE051', '\uE052', '\uE053',
+            '\uE054', '\uE055', '\uE056', '\uE057', '\uE058', '\uE059', '\uE05A', '\uE05B', '\uE05C', '\uE05D',
+            '\uE05E', '\uE05F', '\uE060', '\uE061', '\uE062', '\uE063', '\uE064', '\uE065', '\uE066', '\uE067',
+            '\uE068', '\uE069', '\uE06A', '\uE06B', '\uE06C', '\uE06D', '\uE06E', '\uE06F');
```
Comment (ShadowCat117) [inline common/src/main/java/com/wynntils/models/wynnfont/WynnFontModel.java]: If you make this just the characters that only work in the fancy font, you can populate both maps in the one loop and then add these in a separate one. And then you'll need a separate normal characters list too

Comment (Rafii2198) [reply common/src/main/java/com/wynntils/models/wynnfont/WynnFontModel.java]: Would something like this work?
```java
        for (int i = 0; i < fancyCharacters.size(); i++) {
            normalToFancy.put(normalCharacters.get(i), fancyCharacters.get(i));
            if (i < backgroundCharacters.size()) normalToBackground.put(normalCharacters.get(i), backgroundCharacters.get(i));
        }

Comment (ShadowCat117) [reply common/src/main/java/com/wynntils/models/wynnfont/WynnFontModel.java]: The issue I'm trying to solve is that both lists are currently the same besides the 6 characters that only work for fancy meaning if in the unlikely event we have to change the values, they need to be changed in both which I'm sure we would remember to do it's just better to not even have the risk

Comment (Rafii2198) [reply common/src/main/java/com/wynntils/models/wynnfont/WynnFontModel.java]: That should be it then

**Review by ShadowCat117 — APPROVED** [review]
Looks good now! Excited for this one to get released

**Review by kristofbolyai — APPROVED** [review]
